lib/Conversion/PipelineToHW/PipelineToHW.cpp

```diff
@@ -225,21 +246,24 @@ class PipelineOutlineLowering : public PipelineLowering {
   // Helper class to manage grabbing the various inputs for stage modules.
   struct PipelineStageMod {
     PipelineStageMod(PipelineOutlineLowering &parent, Block *block,
-                     hw::HWModuleOp mod) {
+                     hw::HWModuleOp mod, bool withStall) {
       size_t nStageInputArgs = parent.pipeline.getNumStageArguments(block);
       inputs = mod.getArguments().take_front(nStageInputArgs);
       llvm::SmallVector<Value> stageExtInputs;
       parent.getStageExtInputs(block, stageExtInputs);
       extInputs =
           mod.getArguments().slice(nStageInputArgs, stageExtInputs.size());
-      valid = mod.getArgument(mod.getNumArguments() - 3);
+      valid = mod.getArgument(mod.getNumArguments() - (withStall ? 4 : 3));
+      if (withStall)
+        stall = mod.getArgument(mod.getNumArguments() - 3);
       clock = mod.getArgument(mod.getNumArguments() - 2);
       reset = mod.getArgument(mod.getNumArguments() - 1);
```

Comment:
These assumed offsets make this lowering code brittle. Could you please break them out into well-named constants which are used everywhere in the lower in a subsequent "cleanup" PR?

Reply:
like 'resetOffsetFromEnd'?

And maybe add an assert that it's the expected value or type in places where you assemble an array?

Whenever I see magic numbers, I get concerned that future modifications will break this in a way not caught by a test.